In [404]:
#to find some kind of geogrphical pattern for a person in some are
#source:CDR Data

In [405]:
import pandas as pd
import numpy as np

In [406]:
data = pd.read_csv("tryit.csv", sep = ",")
#print data.head()
#eliminating useless information
data=data.drop(['OTHER','PHONE','TYPE','DURATION','IMEINUMBER','OTYPE','AZIMUTH','OPERATOR','CELLTOWERID','STATE','LAT','LONG'],axis=1)
print data.columns
print data.head()

Index([u'NICKNAME', u'STARTDATE', u'STARTTIME', u'AREADESCRIPTION'], dtype='object')
  NICKNAME   STARTDATE   STARTTIME                  AREADESCRIPTION
0   Jackal  01/01/2014  5:40:06 AM  BSC80:20_MHBD_AnuComplx_WL2033A
1   Jackal  01/01/2014  5:51:43 AM  BSC80:20_MHBD_AnuComplx_WL2033A
2   Jackal  01/01/2014  7:13:29 AM  BSC80:20_MHBD_AnuComplx_WL2033A
3   Jackal  01/01/2014  2:31:08 PM  BSC80:20_MHBD_AnuComplx_WL2033A
4   Jackal  01/01/2014  2:40:11 PM  BSC80:20_MHBD_AnuComplx_WL2033A


In [407]:
#unique candidates in data
print data.NICKNAME.unique()

['Jackal' 'Red Fox' 'King Pin' 'Blackbeard' 'Peacock']


<h1>Models to predict <i>JACKAL's</i> LOCATION</h1>

In [408]:
#print data[['AREADESCRIPTION','LAT','LONG']].drop_duplicates()
mew=data.loc[data['NICKNAME']=='Jackal'].loc[data['AREADESCRIPTION']!='NOT AVAILABLE']
print "distinct Location visited by Jackal::\n",location_jack.head()

distinct Location visited by Jackal::
                    AREADESCRIPTION        LAT       LONG
0   BSC80:20_MHBD_AnuComplx_WL2033A  17.603528  80.004778
8         BSC80:33_Parakala_WL2039A  18.194083  79.701111
11  BSC80:57_MHBD_GovtHsptl_WL2079A  17.594444  80.001694
12   BSC80:21_MHBD_TeleExge_WL2003A  17.596361  79.995222
18   BSC01:32_Apparajupally_WL2138B    17.7142     79.995


In [409]:
from datetime import date
import calendar
import re
mew['TIME']='__'
mew['DAY']='__'
#print mew
for index, row in mew.iterrows():
    #print row['STARTDATE']
    d,m,y=map(int,row['STARTDATE'].split('/'))
    #print date(y,m,d).weekday()
    row['DAY']=date(y,m,d).weekday()
    g=re.split(':| ',row['STARTTIME'])
    if g[3]=='AM' and g[0]!='12':
        row['TIME']=int(g[0])
        
    elif g[3]=='AM' and g[0]=='12':
        row['TIME']=0
        
    elif g[3]=='PM' and g[0]=='12':
        row['TIME']=12
        
    else:
        row['TIME']=int(g[0])+12        
    


In [410]:
print mew[:15]

   NICKNAME   STARTDATE    STARTTIME                  AREADESCRIPTION TIME DAY
0    Jackal  01/01/2014   5:40:06 AM  BSC80:20_MHBD_AnuComplx_WL2033A    5   2
1    Jackal  01/01/2014   5:51:43 AM  BSC80:20_MHBD_AnuComplx_WL2033A    5   2
2    Jackal  01/01/2014   7:13:29 AM  BSC80:20_MHBD_AnuComplx_WL2033A    7   2
3    Jackal  01/01/2014   2:31:08 PM  BSC80:20_MHBD_AnuComplx_WL2033A   14   2
4    Jackal  01/01/2014   2:40:11 PM  BSC80:20_MHBD_AnuComplx_WL2033A   14   2
5    Jackal  01/01/2014   7:26:21 PM  BSC80:20_MHBD_AnuComplx_WL2033A   19   2
6    Jackal  02/01/2014   6:34:54 AM  BSC80:20_MHBD_AnuComplx_WL2033A    6   3
7    Jackal  02/01/2014   8:00:50 AM  BSC80:20_MHBD_AnuComplx_WL2033A    8   3
8    Jackal  02/01/2014   7:46:02 PM        BSC80:33_Parakala_WL2039A   19   3
9    Jackal  02/01/2014   7:47:57 PM        BSC80:33_Parakala_WL2039A   19   3
10   Jackal  03/01/2014   7:39:20 AM  BSC80:20_MHBD_AnuComplx_WL2033A    7   4
11   Jackal  03/01/2014  10:33:46 AM  BSC80:57_MHBD_

In [411]:
mew=mew.drop(['STARTDATE','STARTTIME','NICKNAME'],axis=1)
mew=mew[['DAY','TIME','AREADESCRIPTION']]

In [477]:
#deleted additional features....... like date,time bcoz we already got genaralised 
print mew.head()

  DAY TIME  AREADESCRIPTION
0   2    5               26
1   2    5               26
2   2    7               26
3   2   14               26
4   2   14               26


In [413]:
from IPython.core.display import HTML

<h1>done with preprocessing </h1>

In [414]:
print mew['AREADESCRIPTION'].value_counts()

BSC80:20_MHBD_AnuComplx_WL2033A       324
BSC80:57_MHBD_GovtHsptl_WL2079A        64
BSC80:21_MHBD_TeleExge_WL2003A         20
BSC77:09_TNGOBLDG_WL2114A              14
BSC80:38_Kothaguda_WL2076A             14
BSC80:26_Gudur_WL2028A                 13
BSC80:51_Narsampet2_WL2084A            13
BSC80:28_NarsampetaTE_WL2004A          11
BSC77:18_Tadwai_HW_WL2045A              9
BSC77:20_Medaram_EM_WL2251B             9
BSC77:23_Tirumalagandi_WL2135B          9
BSC77:02_Kambhalapalle_WL2168B          8
BSC80:17_Mulugu_TE_WL2035A              8
BSC80:36_Khanapur_WL2074A               6
BSC01:32_Apparajupally_WL2138B          6
BSC80:08_Pasra_WL2041A                  4
BSC80:56_MondraiH2M_WL2264C             4
BSC01:33_Ootai_WL2198A                  4
BSC77:14_EturnagaramBS_WL2228A          3
BSC77:28_Kesamudram_vlg_WL2136C         2
BSC72:28_Laxettpet_AD2014A              2
BSC80:59_Mallampalle_WL2257B            2
BSC77:25_MDRM_Jakaram_WL2274C           2
BSC77:19_Narlapur_WL2252A         

In [415]:
from sklearn.preprocessing import LabelEncoder

lb_make = LabelEncoder()

mew['AREADESCRIPTION']=lb_make.fit_transform(mew['AREADESCRIPTION'])
print mew.head()

X=mew[['DAY','TIME']].values
#print type(X)
y=mew['AREADESCRIPTION'].values
#print type(y)

  DAY TIME  AREADESCRIPTION
0   2    5               26
1   2    5               26
2   2    7               26
3   2   14               26
4   2   14               26


<h1>Lets start with Support Vector Machine<h2>


In [416]:
from sklearn import svm
clf=svm.SVC(gamma=0.05, C=15)

In [417]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [418]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(clf,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

Cross Validation Score : 69.3495679569 %


In [419]:
clf.fit(X_train,y_train)
print "Test score,",clf.score(X_test,y_test),"%"

Test score, 0.620689655172 %


<h1>Example Prediction Using SVM</h1>

In [420]:
import numpy
print "example prediction:Jackal's position on Monday,7 pm"
a=numpy.asarray([1,19])
print lb_make.inverse_transform(clf.predict(a))[0]
print "\n\n"

print "example prediction:Jackal's position on Thursday,2 pm"
a=numpy.asarray([4,14])
print lb_make.inverse_transform(clf.predict(a))[0]

example prediction:Jackal's position on Monday,7 pm
BSC80:20_MHBD_AnuComplx_WL2033A



example prediction:Jackal's position on Thursday,2 pm
BSC80:57_MHBD_GovtHsptl_WL2079A


/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


<h1>Lets try Naive Bayes</h1>

In [458]:
from sklearn.naive_bayes import GaussianNB
gnb = GaussianNB()

In [459]:
gnb.fit(X_train,y_train)

GaussianNB(priors=None)

In [491]:
print "Prediction Accuracy, ",gnb.score(X_test,y_test)*100,"%"

scores = cross_val_score(gnb,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

Prediction Accuracy,  50.0 %
Cross Validation Score : 48.7337960442 %


<h1>Lets try KNN</h1>

In [492]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=35)
knn.fit(X_train,y_train)
print "Prediction Accuracy, ",knn.score(X_test,y_test)*100,"%"


scores = cross_val_score(knn,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

Prediction Accuracy,  65.5172413793 %
Cross Validation Score : 68.7114625637 %


In [478]:
print "example prediction:Jackal's position on Monday,7 pm"
a=numpy.asarray([1,19])
print lb_make.inverse_transform(knn.predict(a))[0]
print "\n\n"

print "example prediction:Jackal's position on Thursday,2 pm"
a=numpy.asarray([4,14])
print lb_make.inverse_transform(knn.predict(a))[0]

example prediction:Jackal's position on Monday,7 pm
BSC80:20_MHBD_AnuComplx_WL2033A



example prediction:Jackal's position on Thursday,2 pm
BSC80:57_MHBD_GovtHsptl_WL2079A


/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)
/home/siddharth/miniconda2/lib/python2.7/site-packages/sklearn/utils/validation.py:395: DeprecationWarning: Passing 1d arrays as data is deprecated in 0.17 and will raise ValueError in 0.19. Reshape your data either using X.reshape(-1, 1) if your data has a single feature or X.reshape(1, -1) if it contains a single sample.
  DeprecationWarning)


Both KNN and SVM have predicted the same value for trial

In [485]:
from sklearn import linear_model
logistic = linear_model.LogisticRegression(C=5000)
logistic.fit(X_train,y_train)

LogisticRegression(C=5000, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [496]:
print "prediction accuracy",logistic.score(X_test,y_test)*100," %"

scores = cross_val_score(logistic,X_train,y_train, cv=27)
print "Cross Validation Score :",scores.mean()*100,"%"

prediction accuracy 58.6206896552  %
Cross Validation Score : 64.845082409 %


<h1><center>FINAL RESULT</center><br>
</h1>
SVM=>test:62% cross validation:69.34%
<br>KNN=>test:65% cross validation:69%
<br>Logistic Regression=>test:58% cross validation:65%
<br>Naive Bayes=>test:50% cross validation:48%


KNN has the highest test accuracy and good cross validation Score.